File: binding/src/libqcollator.cpp
#include <QChar>
#include <QCollator>
#include <QCollatorSortKey>
#include <QLocale>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QStringView>
#include <qcollator.h>
#include "libqcollator.h"
#include "libqcollator.hxx"

QCollatorSortKey* QCollatorSortKey_new(const QCollatorSortKey* other) {
	 return new QCollatorSortKey(*other);
}

void QCollatorSortKey_OperatorAssign(QCollatorSortKey* self, const QCollatorSortKey* other) {
	self->operator=(*other);
}

void QCollatorSortKey_Swap(QCollatorSortKey* self, QCollatorSortKey* other) {
	self->swap(*other);
}

int QCollatorSortKey_Compare(const QCollatorSortKey* self, const QCollatorSortKey* key) {
	return self->compare(*key);
}

void QCollatorSortKey_Delete(QCollatorSortKey* self) {
    delete self;
}

QCollator* QCollator_new() {
	 return new QCollator();
}

QCollator* QCollator_new2(const QLocale* locale) {
	 return new QCollator(*locale);
}

QCollator* QCollator_new3(const QCollator* param1) {
	 return new QCollator(*param1);
}

void QCollator_OperatorAssign(QCollator* self, const QCollator* param1) {
	self->operator=(*param1);
}

void QCollator_Swap(QCollator* self, QCollator* other) {
	self->swap(*other);
}

void QCollator_SetLocale(QCollator* self, const QLocale* locale) {
	self->setLocale(*locale);
}

QLocale* QCollator_Locale(const QCollator* self) {
	return new QLocale(self->locale());
}

int QCollator_CaseSensitivity(const QCollator* self) {
	return self->caseSensitivity();
}

void QCollator_SetCaseSensitivity(QCollator* self, int cs) {
	self->setCaseSensitivity(static_cast<Qt::CaseSensitivity>(cs));
}

void QCollator_SetNumericMode(QCollator* self, bool on) {
	self->setNumericMode(on);
}

bool QCollator_NumericMode(const QCollator* self) {
	return self->numericMode();
}

void QCollator_SetIgnorePunctuation(QCollator* self, bool on) {
	self->setIgnorePunctuation(on);
}

bool QCollator_IgnorePunctuation(const QCollator* self) {
	return self->ignorePunctuation();
}

int QCollator_Compare(const QCollator* self, const libqt_string s1, const libqt_string s2) {
	return self->compare(QString::fromUtf8(s1.data, s1.len), QString::fromUtf8(s2.data, s2.len));
}

int QCollator_Compare2(const QCollator* self, const QChar* s1, ptrdiff_t len1, const QChar* s2, ptrdiff_t len2) {
	return self->compare(s1, len1, s2, len2);
}

bool QCollator_OperatorCall(const QCollator* self, const libqt_string s1, const libqt_string s2) {
	return self->operator()(QString::fromUtf8(s1.data, s1.len), QString::fromUtf8(s2.data, s2.len));
}

int QCollator_Compare3(const QCollator* self, QStringView* s1, QStringView* s2) {
	return self->compare(*s1, *s2);
}

bool QCollator_OperatorCall2(const QCollator* self, QStringView* s1, QStringView* s2) {
	return self->operator()(*s1, *s2);
}

QCollatorSortKey* QCollator_SortKey(const QCollator* self, const libqt_string stringVal) {
	return new QCollatorSortKey(self->sortKey(QString::fromUtf8(stringVal.data, stringVal.len)));
}

int QCollator_DefaultCompare(QStringView* s1, QStringView* s2) {
	return QCollator::defaultCompare(*s1, *s2);
}

QCollatorSortKey* QCollator_DefaultSortKey(QStringView* key) {
	return new QCollatorSortKey(QCollator::defaultSortKey(*key));
}

void QCollator_Delete(QCollator* self) {
    delete self;
}

File: binding/src/libqcollator.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qcollatorsortkey.html
pub const qcollatorsortkey = struct {

    /// New constructs a new QCollatorSortKey object.
    pub fn New(other: ?*anyopaque) QtC.QCollatorSortKey {
        return qtc.QCollatorSortKey_new(@ptrCast(other));
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QCollatorSortKey_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QCollatorSortKey_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn Compare(self: ?*anyopaque, key: ?*anyopaque) i32 {
        return qtc.QCollatorSortKey_Compare(@ptrCast(self), @ptrCast(key));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QCollatorSortKey_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qcollator.html
pub const qcollator = struct {

    /// New constructs a new QCollator object.
    pub fn New() QtC.QCollator {
        return qtc.QCollator_new();
    }


    /// New2 constructs a new QCollator object.
    pub fn New2(locale: ?*anyopaque) QtC.QCollator {
        return qtc.QCollator_new2(@ptrCast(locale));
    }


    /// New3 constructs a new QCollator object.
    pub fn New3(param1: ?*anyopaque) QtC.QCollator {
        return qtc.QCollator_new3(@ptrCast(param1));
    }


    pub fn OperatorAssign(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QCollator_OperatorAssign(@ptrCast(self), @ptrCast(param1));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QCollator_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn SetLocale(self: ?*anyopaque, locale: ?*anyopaque) void {
        qtc.QCollator_SetLocale(@ptrCast(self), @ptrCast(locale));
    }

    pub fn Locale(self: ?*anyopaque, ) QtC.QLocale {
        return qtc.QCollator_Locale(@ptrCast(self));
    }

    pub fn CaseSensitivity(self: ?*anyopaque, ) i32 {
        return qtc.QCollator_CaseSensitivity(@ptrCast(self));
    }

    pub fn SetCaseSensitivity(self: ?*anyopaque, cs: i32) void {
        qtc.QCollator_SetCaseSensitivity(@ptrCast(self), @intCast(cs));
    }

    pub fn SetNumericMode(self: ?*anyopaque, on: bool) void {
        qtc.QCollator_SetNumericMode(@ptrCast(self), on);
    }

    pub fn NumericMode(self: ?*anyopaque, ) bool {
        return qtc.QCollator_NumericMode(@ptrCast(self));
    }

    pub fn SetIgnorePunctuation(self: ?*anyopaque, on: bool) void {
        qtc.QCollator_SetIgnorePunctuation(@ptrCast(self), on);
    }

    pub fn IgnorePunctuation(self: ?*anyopaque, ) bool {
        return qtc.QCollator_IgnorePunctuation(@ptrCast(self));
    }

    pub fn Compare(self: ?*anyopaque, s1: []const u8, s2: []const u8) i32 {
        const s1_str = qtc.libqt_string{
    .len = s1.len,
    .data = s1.ptr,
};
const s2_str = qtc.libqt_string{
    .len = s2.len,
    .data = s2.ptr,
};
return qtc.QCollator_Compare(@ptrCast(self), s1_str, s2_str);
    }

    pub fn Compare2(self: ?*anyopaque, s1: ?*anyopaque, len1: i64, s2: ?*anyopaque, len2: i64) i32 {
        return qtc.QCollator_Compare2(@ptrCast(self), @ptrCast(s1), len1, @ptrCast(s2), len2);
    }

    pub fn OperatorCall(self: ?*anyopaque, s1: []const u8, s2: []const u8) bool {
        const s1_str = qtc.libqt_string{
    .len = s1.len,
    .data = s1.ptr,
};
const s2_str = qtc.libqt_string{
    .len = s2.len,
    .data = s2.ptr,
};
return qtc.QCollator_OperatorCall(@ptrCast(self), s1_str, s2_str);
    }

    pub fn Compare3(self: ?*anyopaque, s1: []const u8, s2: []const u8) i32 {
        return qtc.QCollator_Compare3(@ptrCast(self), @ptrCast(s1), @ptrCast(s2));
    }

    pub fn OperatorCall2(self: ?*anyopaque, s1: []const u8, s2: []const u8) bool {
        return qtc.QCollator_OperatorCall2(@ptrCast(self), @ptrCast(s1), @ptrCast(s2));
    }

    pub fn SortKey(self: ?*anyopaque, stringVal: []const u8) QtC.QCollatorSortKey {
        const stringVal_str = qtc.libqt_string{
    .len = stringVal.len,
    .data = stringVal.ptr,
};
return qtc.QCollator_SortKey(@ptrCast(self), stringVal_str);
    }

    pub fn DefaultCompare(s1: []const u8, s2: []const u8) i32 {
        return qtc.QCollator_DefaultCompare(@ptrCast(s1), @ptrCast(s2));
    }

    pub fn DefaultSortKey(key: []const u8) QtC.QCollatorSortKey {
        return qtc.QCollator_DefaultSortKey(@ptrCast(key));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QCollator_Delete(@ptrCast(self));
    }
};
